### ISSUE TITLE
Performance degradation in 10.4.1 `arrayHasKey` with specific inputs

### ISSUE THREAD

**Issue by dktapps** [sebastianbergmann/phpunit#5539]
| Q                   | A
| --------------------| ---------------
| PHPUnit version     | 10.4.1
| PHP version         | 8.2.9
| Installation Method | Composer

#### Summary
With specific inputs I'm noticing that `arrayHasKey` takes many times longer with `10.4.0` and upwards than it did with `10.3.4` with certain inputs.

#### Current behavior
arrayHasKey is much slower, causing a massive slowdown in tests using arrayHasKey many thousands of times. My unchanged test suite now takes upwards of 45 seconds, where before it ran in about 2 seconds. Profiling with xdebug suggests that `sebastian/exporter` may be the culprit, although I'm not really sure why.

![image](https://github.com/sebastianbergmann/phpunit/assets/14214667/98fb0e42-8218-484c-ae60-cbfe7445e132)

#### How to reproduce
Use the following test case:
```php
<?php

declare(strict_types=1);

use PHPUnit\Framework\TestCase;
use function json_decode;
use const JSON_THROW_ON_ERROR;

final class SlowTest extends TestCase{

	public function testSlowTest() : void{
		$knownStates = json_decode(file_get_contents(__DIR__ . "/phpunit-slow-test.json"), true, JSON_THROW_ON_ERROR);

		foreach($knownStates as $id => $name){
			self::assertArrayHasKey($id, $knownStates);
		}
	}
}
```
with the following data:
[phpunit-slow-test.json](https://github.com/sebastianbergmann/phpunit/files/12921409/phpunit-slow-test.json)

With 10.3.x, this test runs almost instantly. With 10.4.x, the test takes almost 45 seconds to complete.

#### Expected behavior
The test should run almost immediately.

**Comment by dktapps** [comment]
This issue appears to still be present as of 10.5.3

**Comment by sebastianbergmann** [comment]
"appears to still be present" does not sound like you know for sure. Can you please make sure? Thanks! 

Also may I ask how big the array in question is and whether it only contains scalar values?

**Comment by dktapps** [comment]
> Also may I ask how big the array in question is and whether it only contains scalar values?

I linked a sample in the issue. It's a large array of string => string, and contains probably around 5000 items or so.

**Comment by sebastianbergmann** [comment]
I can confirm that this issue is still present in PHPUnit 10.5:

#### PHPUnit 10.5
```
PHPUnit 10.5.8-4-gbd134ad32a by Sebastian Bergmann and contributors.

Runtime:       PHP 8.3.2

.                                                                   1 / 1 (100%)

Time: 01:53.979, Memory: 6.00 MB

OK (1 test, 7942 assertions)
```

I can also confirm that this issue is no longer present in PHPUnit 11.0:

#### PHPUnit 11.0
```
PHPUnit 11.0-gedf7f2bd74 by Sebastian Bergmann and contributors.

Runtime:       PHP 8.3.2

.                                                                   1 / 1 (100%)

Time: 00:00.014, Memory: 6.00 MB

OK (1 test, 7942 assertions)
```
A fix for this performance issue cannot be implemented in PHPUnit 10.5 without breaking backward compatibility.

**Comment by dktapps** [comment]
Thanks! What's the timeline for 11.0.0 release?
